File: ruby/string_compressor.rb
# A method to compress a string of alphabetical characters.
# Eg: AAAABBBGGGEEE would become A4B3G3E3
# if the compressed string will be longer than the original
# string, return the original string

def compress_string(str)
  # compressed string will be the uniqued string * 2 because of the count
  comp_length = str.chars.uniq.length * 2

  # check if compressed string will be shorter than input string
  if comp_length < str.length
    compressed_string = ""
    count = 1
    # iterate over the string
    str.each_char.with_index do |letter, index|
      # if following letter in string is same, increment the count
      #  else add letter and count to compressed_string var
      if letter == str[index+1]
        count += 1
      else
        compressed_string << letter
        compressed_string << count.to_s
        count = 1
      end
    end
    compressed_string
  else
    str
  end
end

File: javascript/string_compressor.js
// A method to compress a string of alphabetical characters. 
// Eg: AAAABBBGGGEEE would become A4B3G3E3

function stringCompressor(str){
  var compressed_string = "";
  var count = 1;
  for(var i = 0; i<str.length; i++){
    if(str[i] == str[i+1]){
      count++
    } else {
      compressed_string += str[i];
      compressed_string += count.toString();
      count = 1;
    }
  }
  return compressed_string;
}
